### PR TITLE
Resilience experiments with Rayleigh-Benard convection

### PR DESCRIPTION
**Pull request by brownbaerchen**
This PR includes a configuration for Rayleigh-Benard convection (RBC) on a 256x128 grid with Rayleigh number of $2^4$. Reference solutions are computed for t=20 and t=21. Individual runs for fault experiments then start at t=20 and go to t=21. There is a convergence controller that makes sure the target time is reached to accuracy of $10^{-9}$, which is sufficiently accurate to use the reference solution. I verified that I get the order of the collocation problem relative to such a reference solution, even if the residual stays at order 1 during the runs. The latest theory is that this is because the problem is perturbed by the method that enforces the boundary conditions. This perturbation is missing in the residual and I haven't found a good way to represent it there (yet). This matters when the perturbation is not small, i.e. the resolution is insufficient in z-direction. The configuration used here, however, is sufficiently resolved for this not to cause any issues.

### CODE REVIEW

**Comment by pancetta** [comment]
Hm, is anything of this tested somewhere?

**Comment by brownbaerchen** [comment]
No, that's too expensive, I'm afraid. I had to choose a large resolution to get low residuals in the collocation problem and now these experiments take about 40s+ on my laptop each. Also, they need the reference solution. Computing that takes multiple hours, and checking it in is not great either.

**Comment by brownbaerchen** [comment]
When doing experiments with work-precision plots, I noticed some issues with the reference solution. These are fixed now, there is now a test that the time that is supposed to be reached is actually reached very closely. The result is also more sensitivity to faults in RBC.
Also, I implemented a convergence controller that rounds the step size. The idea is that the cache of matrix factorisations gets a higher hit rate if there are fewer options for step sizes. I.e. if adaptivity wants a step size of 0.612412, but you have a factorisation for 0.6 in the cache, then it does not need to be recomputed if the step size is rounded to that value. This is also tested.
The parameters for the work-precision plot with RBC are not final, but since this includes some tests, I thought I add this already to this PR and change again in the future.

File: pySDC/projects/Resilience/RBC.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    return data
+
+
+RayleighBenard._u_exact = RayleighBenard.u_exact
+RayleighBenard.u_exact = u_exact
+
+PROBLEM_PARAMS = {'Rayleigh': 2e4, 'nx': 256, 'nz': 128}
+
+
+class ReachTendExactly(ConvergenceController):
```
Comment (pancetta) [inline pySDC/projects/Resilience/RBC.py:63]: No documentation here?

File: pySDC/projects/Resilience/RBC.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    ax.loglog(dts, errors, label=f'{num_nodes} nodes', marker='x')
+    ax.loglog(
+        dts, [max_error * (me / dts[0]) ** (2 * num_nodes - 1) for me in dts], ls='--', label=f'order {2*num_nodes-1}'
+    )
+    ax.set_xlabel(r'$\Delta t$')
+    ax.set_ylabel(r'global error')
+    ax.legend(frameon=False)
+
+
+def test_order(t=14, dt=1e-1, steps=6):
```
Comment (pancetta) [inline pySDC/projects/Resilience/RBC.py:278]: Sure you want to call this `test_xxx` ?

**Pull request merged**